primitives/trie/src/cache/mod.rs

```diff
@@ -36,43 +36,285 @@
+	/// The current size (in bytes) of data allocated by this cache on the heap.
+	///
+	/// This doesn't include the size of the map itself.
+	current_heap_size: usize,
+}
+
+impl<H> schnellru::Limiter<H, NodeCached<H>> for LocalNodeCacheLimiter
+where
+	H: AsRef<[u8]> + std::fmt::Debug,
+{
+	type KeyToInsert<'a> = H;
+	type LinkType = u32;
+
+	#[inline]
+	fn is_over_the_limit(&self, length: usize) -> bool {
+		// Only enforce the limit if there's more than one element to make sure
+		// we can always add a new element to the cache.
+		if length <= 1 {
+			return false
 		}
+
+		self.current_heap_size > LOCAL_NODE_CACHE_MAX_HEAP_SIZE
+	}
+
+	#[inline]
+	fn on_insert<'a>(
+		&mut self,
+		_length: usize,
+		key: H,
+		cached_node: NodeCached<H>,
+	) -> Option<(H, NodeCached<H>)> {
+		self.current_heap_size += cached_node.heap_size();
+		Some((key, cached_node))
+	}
+
+	#[inline]
+	fn on_replace(
+		&mut self,
+		_length: usize,
+		_old_key: &mut H,
+		_new_key: H,
+		old_node: &mut NodeCached<H>,
+		new_node: &mut NodeCached<H>,
+	) -> bool {
+		debug_assert_eq!(_old_key.as_ref().len(), _new_key.as_ref().len());
+		self.current_heap_size =
+			self.current_heap_size + new_node.heap_size() - old_node.heap_size();
+		true
+	}
+
+	#[inline]
+	fn on_removed(&mut self, _key: &mut H, cached_node: &mut NodeCached<H>) {
+		self.current_heap_size -= cached_node.heap_size();
+	}
+
+	#[inline]
+	fn on_cleared(&mut self) {
+		self.current_heap_size = 0;
+	}
+
+	#[inline]
+	fn on_grow(&mut self, new_memory_usage: usize) -> bool {
+		new_memory_usage <= LOCAL_NODE_CACHE_MAX_INLINE_SIZE
+	}
+}
+
+/// A limiter for the local value cache. This makes sure the local cache doesn't grow too big.
+#[derive(Default)]
+pub struct LocalValueCacheLimiter {
+	/// The current size (in bytes) of data allocated by this cache on the heap.
+	///
+	/// This doesn't include the size of the map itself.
+	current_heap_size: usize,
+}
+
+impl<H> schnellru::Limiter<ValueCacheKey<H>, CachedValue<H>> for LocalValueCacheLimiter
+where
+	H: AsRef<[u8]>,
+{
+	type KeyToInsert<'a> = ValueCacheRef<'a, H>;
+	type LinkType = u32;
+
+	#[inline]
+	fn is_over_the_limit(&self, length: usize) -> bool {
+		// Only enforce the limit if there's more than one element to make sure
+		// we can always add a new element to the cache.
+		if length <= 1 {
+			return false
+		}
+
+		self.current_heap_size > LOCAL_VALUE_CACHE_MAX_HEAP_SIZE
+	}
+
+	#[inline]
+	fn on_insert(
+		&mut self,
+		_length: usize,
+		key: Self::KeyToInsert<'_>,
+		value: CachedValue<H>,
+	) -> Option<(ValueCacheKey<H>, CachedValue<H>)> {
+		self.current_heap_size += key.storage_key.len();
+		Some((key.into(), value))
+	}
+
+	#[inline]
+	fn on_replace(
+		&mut self,
+		_length: usize,
+		_old_key: &mut ValueCacheKey<H>,
+		_new_key: ValueCacheRef<H>,
+		_old_value: &mut CachedValue<H>,
+		_new_value: &mut CachedValue<H>,
+	) -> bool {
+		debug_assert_eq!(_old_key.storage_key.len(), _new_key.storage_key.len());
```

Comment:
Yes, they are the same.

Basically, when a new element is inserted and there already exist an element whose key has the same hash and is equal to the new element's key then this method is called. But technically the keys are two different objects, and just because they hash the same and they are equal to each other doesn't necessary mean they are exactly the same, depending on how `Hash` and `PartialEq` are implemented. And in this case `old_key` and `new_key` *are* different types (the existing key inside of the map is owned, while the newly inserted key is borrowed), so the traits must have been implemented manually, so they could be buggy, hence this assertion that makes sure the replaced key's length is the same so that the `current_heap_size` doesn't desync and doesn't start to underflow/overflow.

Reply:
I've [updated the docs](https://github.com/koute/schnellru/commit/b9d5a47d6fae4125ab8433a1cbfc1ce25e4a577f); hopefully it's more clear now. (: